### ISSUE TITLE
Question: signing previous commit

### ISSUE THREAD

**Issue by thisguychris** [newren/git-filter-repo#67]
Thank you for making this tool! I've tried updating the email, it was really fast (useful for emails you don't use anymore).

I stumbled upon this repo when trying to sign my previous commits via git-filter-branch, saying `git-filter-repo` is now the preferred way over `git-filter-branch`

```sh
git filter-branch -f --commit-filter 'git commit-tree -S "$@";' 
```

I was wondering how to achieve this via git-filter-repo? Checking the [Internals section ](https://htmlpreview.github.io/?https://github.com/newren/git-filter-repo/blob/docs/html/git-filter-repo.html) this tool is stripping tags internally? 

My naive question would be signing previous commits in this repo is not possible? If it is, would you mind pointing me to how?

**Comment by newren** [comment]
Ooh, interesting usecase.

fast-export strips signatures on both commits and tags; it has some options around the tags (such as leaving them in place despite them no longer being valid, which filter-repo doesn't use), but has no functionality at all around signatures on commits.  So, you've found the first usecase of something filter-branch can do that filter-repo cannot.

For now, you'll need to continue to use filter-branch.  It may make sense to add additional capabilities to fast-export to sign new commits (and/or tags), add related capability to fast-import to parse it, and then add options to filter-repo to make use of this and plumb everything together.

**Comment by thisguychris** [comment]
Hey @newren thanks for the reply, I'll look out for future features. git-filter-repo is insanely fast! Thanks for making this. 🙏

I tried using mailcap for replacing emails I use, is there an option for changing only a range of commits, instead of the whole history?

My use case is for changing author emails if I use a different GPG key. I don't want to mess up my older commits, which use a separate email and a different GPG key.

**Comment by newren** [comment]
Yes, you can change just a range of commits, but that range cannot exclude any commits that depend upon the ones you rewrite.  For example, if you have a history with 100 commits and a fully linear history (i.e. just one branch, no merges, etc.), then you can rewrite just commits 80-100.  You cannot just rewrite commits 60-80, because changing those commits changes their hashes and the git model makes the hash of any commit dependent on the hashes of its parents.

You may want to take a look at the --refs for doing this (or you could use a --commit-callback and do it based on commit.original_id or commit.committer_date or other such things).

**Comment by thisguychris** [comment]
Thanks, @newren, I understand, since it's a new commit, it'll change hashes and if there's any commit left, it'll mess it's history. I'll definitely check it out. As long as I stay on changing commits at the tip I'll be fine I guess?

**Comment by newren** [comment]
Yes, as long as whatever range of commits you change does not exclude commits that depend on your range, then you're fine.  I'd hesitate to simplify that to "changing commits at the tip" since multiple branches can mean there are multiple tips, but I think you have the right basic idea.

**Comment by catap** [comment]
@newren I don't think that I need to open a new issue, let me explain it here.

Let assume that I have very simple use-case where I'd like to exclude some folder from git repo as dedicated repository. Commit messages contains tonns of cross-references => using `git-filter-repo` seems as very good idea.

Anyway, as soon as commits should be signed, the hell's begin. I can run `git filter-repo --subdirectory-filter articles/` and keep everything inside one folder with rewriting cross references from commit messages.

Anyway, I can't re-sign everything in history because as soon as I signed the first commit, the ID of this commit is changed and it changes all ID for commits that was after this commit => references inside messages is useless.

I can update it by hand and I did it few times before, but here I have 2 years of active development and tons of references — I have no idea how many of them, I guess hundreds.

So, what should I do? Do you have any idea? Thanks.

**Comment by vhscom** [comment]
What OP is looking for may not be possible with `git-filter-repo` see maintainer comment here: https://github.com/newren/git-filter-repo/discussions/209#discussioncomment-474986

**Comment by SmushyTaco** [comment]
I hope this changes in the future. I'm currently using this PowerShell script I made:
```powershell
$originalRemote = git config --get remote.origin.url

@"
Nikan Radan <github.com@alias.nikanradan.com> <smushytaco@gmail.com>
Nikan Radan <github.com@alias.nikanradan.com> <personal@nikanradan.com>
Nikan Radan <github.com@alias.nikanradan.com> <nradan@wgu.edu>
Nikan Radan <github.com@alias.nikanradan.com> <41648290+SmushyTaco@users.noreply.github.com>
Nikan Radan <github.com@alias.nikanradan.com> <organicradan@gmail.com>
"@ | Out-File -FilePath "mailmap" -Encoding utf8

git filter-repo --mailmap mailmap --force

Remove-Item -Path "mailmap"

git filter-branch -f --commit-filter '
if [ "$GIT_AUTHOR_EMAIL" = "github.com@alias.nikanradan.com" ];
then
    git commit-tree -S "$@";
else
    git commit-tree "$@";
fi' -- --all

git remote add origin $originalRemote

$currentBranch = git rev-parse --abbrev-ref HEAD

git push --set-upstream origin $currentBranch --force

```

To update the email and names and sign all my commits but everything is instant until it hits `filter-branch` which takes a way too long.
